File: gold_app/gold_app/page/wholesale_bag_melt/wholesale_bag_melt.js
// gold_app/page/wholesale_bag_melt/wholesale_bag_melt.js

let WBMState = {
	selected_bag: null,
	bag_summary: null,
	bag_items: [],
	bag_list: [],
};

frappe.pages["wholesale-bag-melt"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "Select Wholesale Bag to Melt",
		single_column: true,
	});

	$(page.body).append(`
        <div id="wbdm-root" class="wbdm-page">
            <div class="wbdm-inner">
				<div id="wbdm-loader" class="loader-overlay">
    				<div class="loader"></div>
    				<p>Loading...</p>
				</div>
                <div id="wbdm-bag-list" class="wbdm-grid"></div>
            </div>
        </div>
    `);

	frappe.require("/assets/gold_app/css/wholesale_bag_melt.css");

	$("#wbdm-loader").fadeIn(150);
	$("#wbdm-bag-list").hide();

	frappe.call({
		method: "gold_app.api.sales.wholesale_bag_melt.get_bag_overview",
		callback: function (r) {
			if (r.message) {
				$("#wbdm-loader").fadeOut(200, () => {
					$("#wbdm-bag-list").fadeIn(200);
				});

				WBMState.bag_list = r.message;
				renderBagGrid(WBMState.bag_list);

				WBMState.onBackToBags = function () {
					$("#wbdm-root").html(`
                        <div class="wbdm-inner">
                            <div id="wbdm-bag-list" class="wbdm-grid"></div>
                        </div>
                    `);
					renderBagGrid(WBMState.bag_list);
				};
			}
		},
	});
};

// ==============================================
// RENDER BAG CARD GRID
// ==============================================
function renderBagGrid(bags) {
	const grid = $("#wbdm-bag-list");
	grid.empty();

	const fmtNumber = (val, decimals) => {
		if (val === null || val === undefined || val === "") return "";
		const n = typeof val === "number" ? val : parseFloat(String(val));
		return n.toLocaleString("en-MY", {
			minimumFractionDigits: decimals,
			maximumFractionDigits: decimals,
		});
	};

	bags.forEach((bag) => {
		grid.append(`
            <div class="wbdm-card">
                <div class="wbdm-bag-title">${bag.bag_id}</div>

                <div class="wbdm-row">
                    <span>Total Weight</span>
                    <strong>${fmtNumber(bag.total_weight, 0)}g</strong>
                </div>

                <div class="wbdm-row">
                    <span>XAU (Pure Gold)</span>
                    <strong class="wbdm-blue">${fmtNumber(bag.xau_g, 2)}g</strong>
                </div>

                <div class="wbdm-row">
                    <span>Avg Purity</span>
                    <strong>${fmtNumber(bag.avg_purity, 1)}</strong>
                </div>

                <div class="wbdm-row">
                    <span>Total Cost</span>
                    <strong>RM${fmtNumber(bag.total_cost, 2)}</strong>
                </div>

                <div class="wbdm-row">
                    <span>Cost per Gram</span>
                    <strong>RM${fmtNumber(bag.cost_per_gram, 2)}</strong>
                </div>

                <div class="wbdm-view-items">▼ View Items</div>

                <button class="wbdm-select-btn" data-bag="${bag.bag_id}">
                    SELECT BAG
                </button>
            </div>
        `);
	});

	$(".wbdm-select-btn")
		.off("click")
		.on("click", function () {
			const bagId = $(this).data("bag");

			WBMState.selected_bag = bagId;

			frappe.call({
				method: "gold_app.api.sales.wholesale_bag_melt.get_bag_details",
				args: { bag_id: bagId },
				callback: function (r) {
					if (!r.message) return;

					WBMState.bag_summary = r.message.summary;
					WBMState.bag_items = r.message.items;

					WBMState.bag_summary.record_id = WBMState.selected_bag;
					WBMState.bag_summary.record_date = getTodayDate();

					showBagSummaryUI();
				},
			});
		});

	$(".wbdm-view-items")
		.off("click")
		.on("click", function () {
			const card = $(this).closest(".wbdm-card");
			const bagId = card.find(".wbdm-select-btn").data("bag");

			WBMState.selected_bag = bagId;

			frappe.call({
				method: "gold_app.api.sales.wholesale_bag_melt.get_bag_details",
				args: { bag_id: bagId },
				callback: function (r) {
					if (!r.message) return;

					WBMState.bag_summary = r.message.summary;
					WBMState.bag_items = r.message.items;

					showBagSummaryUI();
				},
			});
		});
}

// ==============================================
// SHOW GLOBAL SHELL + TABS (SAVE BUTTON ADDED)
// ==============================================
function showBagSummaryUI() {
	$("#wbdm-root").html(`
        <div class="wbm-page-shell">

            <!-- HEADER -->
            <div class="wbm-header">
                <div class="wbm-header-inner">

                    <div class="wbm-header-left">
                        <div class="wbm-record-title">Melt & Assay Sales</div>
                        <div class="wbm-record-meta">
                            <a class="wbm-record-id">
                                ${WBMState.bag_summary.record_id || "-"}
                            </a>
                            <span class="wbm-dot-sep">•</span>
                            <span class="wbm-record-date-label">Sale Date:</span>
                            <span class="wbm-record-date-value">
                                ${WBMState.bag_summary.record_date || "-"}
                            </span>
                        </div>
                    </div>

                    <!-- ✅ SAVE BUTTON (RIGHT SIDE) -->
                    <div class="wbm-header-right">
                        <button id="wbm-save-btn" class="wbm-save-btn">Save</button>
                    </div>

                </div>

                <!-- TABS -->
                <div class="wbm-tabs-bar">
                    <div class="wbm-tab-item" data-tab="bag_summary">Bag Summary</div>
                    <div class="wbm-tab-item" data-tab="melting_assay">Melting & Assay</div>
                    <div class="wbm-tab-item" data-tab="buyer_sale">Sales Detail</div>
                    <div class="wbm-tab-item" data-tab="metrics">Metrics</div>
                </div>
            </div>

            <!-- TAB CONTENT -->
            <div class="wbm-page-body">
				<div id="wbm-tab-loader" class="loader-overlay">
    				<div class="loader"></div>
    				<p>Loading...</p>
				</div>
                <div id="wbd-content" class="wbm-main-card"></div>
            </div>

        </div>
    `);

	WBMState.onSaveRecord = saveMeltAssaySales;

	// default tab
	loadTabContent("bag_summary");
	$(`.wbm-tab-item[data-tab="bag_summary"]`).addClass("wbm-tab-active");

	$(".wbm-tab-item").on("click", function () {
		$(".wbm-tab-item").removeClass("wbm-tab-active");
		$(this).addClass("wbm-tab-active");

		const tab = $(this).data("tab");
		loadTabContent(tab);
	});

	$(document).on("click", "#wbm-save-btn", function () {
		if (WBMState.onSaveRecord) {
			WBMState.onSaveRecord();
		}
	});
}

// ==============================================
// TAB CONTENT LOADER
// ==============================================
function loadTabContent(tabName) {
	const map = {
		bag_summary: "/assets/gold_app/js/bag_summary.js",
		melting_assay: "/assets/gold_app/js/melting_assay.js",
		buyer_sale: "/assets/gold_app/js/buyer_sale.js",
		metrics: "/assets/gold_app/js/metrics.js",
	};

	const path = map[tabName];
	if (!path) return;

	// SHOW LOADER
	$("#wbm-tab-loader").fadeIn(150);
	$("#wbd-content").hide();

	frappe.require(path, () => {
		if (window.WBMComponents && window.WBMComponents[tabName]) {
			window.WBMComponents[tabName]($("#wbd-content"), WBMState);

			// HIDE LOADER AFTER RENDER
			$("#wbm-tab-loader").fadeOut(200, () => {
				$("#wbd-content").fadeIn(200);
			});
		}
	});
}

function getTodayDate() {
	const d = new Date();
	return d.toLocaleDateString("en-GB"); // DD/MM/YYYY
}

function saveMeltAssaySales() {
	// 1️⃣ --- BAG SUMMARY DATA ---
	const summary = WBMState.bag_summary || {};
	const bagContents = WBMState.bag_items || [];

	// 2️⃣ --- MELTING & ASSAY DATA ---
	const melt = WBMState.melting || {};
	const assay = WBMState.assay || {};

	// 3️⃣ --- BUYER SALE DATA ---
	const sale = WBMState.sale || {};

	// 4️⃣ --- METRICS DATA ---
	const metrics = WBMState.metrics || {};

	// 5️⃣ --- BUILD CLEAN PAYLOAD ---
	const payload = {
		header: {
			// BAG SUMMARY
			total_weight: summary.total_weight_g,
			avg_purity: summary.average_purity,
			total_xau: summary.pure_gold_xau_g,
			total_cost: summary.total_cost_basis,
			xau_avco:
				summary.total_cost_basis && summary.pure_gold_xau_g
					? summary.total_cost_basis / summary.pure_gold_xau_g
					: 0,

			// MELTING
			weight_before_melting: melt.before,
			weight_after_melting: melt.after,
			melting_cost: melt.cost,
			melting_payment_mode: melt.payment_mode,
			weight_loss: melt.weight_loss,
			xau_loss: melt.xau_loss,
			loss_percentage: melt.loss_percentage,

			// ASSAY
			current_avg_purity: assay.current_purity,
			assay_purity: assay.assay_purity,
			purity_variance: assay.purity_variance,
			xau_weight_variance: assay.xau_weight_variance,
			actual_xau_weight: assay.actual_xau_weight,
			assay_sample_weight: assay.sample_weight,
			net_xau_sellable: assay.net_sellable,
			assay_cost: assay.cost,
			assay_payment_mode: assay.payment_mode,

			// SALES DATA
			sale_net_weight: sale.net_weight,
			sale_assay_purity: sale.assay_purity,
			sale_net_xau: sale.net_xau,
			total_xau_sold: sale.total_xau_sold,
			total_revenue: sale.revenue,
			weighted_avg_rate: sale.weighted_avg_rate,

			// METRICS (optional)
			...metrics,
		},

		bag_contents: bagContents.map((r) => ({
			purity: r.purity,
			weight: r.weight_g,
			avco: r.cost_per_g_rm,
			cost: r.cost_rm,
			xau: r.xau_g,
			xau_avco: r.xau_g ? r.cost_rm / r.xau_g : 0,
		})),

		locked_rates: sale.locked_rates || [],
	};

	// 6️⃣ --- CALL BACKEND API ---
	frappe.call({
		method: "gold_app.api.sales.wholesale_bag_melt.save_melt_assay_sales",
		args: { payload },
		freeze: true,
		callback: function (r) {
			frappe.msgprint("Melt & Assay Sales saved successfully.");
		},
	});
}

File: gold_app/api/sales/wholesale_bag_melt.py
import frappe
from frappe import _
from frappe.utils import flt, nowdate

@frappe.whitelist()
def get_bag_overview():
    """
    Returns dynamic bag overview for UI (main grid)
    Includes:
        - total_weight
        - avg_purity
        - total_cost
        - cost_per_gram
        - xau_g (calculated pure gold)
    """

    rows = frappe.db.sql("""
        SELECT
            bin.warehouse AS bag_id,
            item.purity AS purity,
            SUM(bin.actual_qty) AS weight,
            AVG(bin.valuation_rate) AS rate,
            SUM(bin.actual_qty * bin.valuation_rate) AS amount
        FROM `tabBin` bin
        JOIN `tabItem` item
            ON item.name = bin.item_code
        JOIN `tabWarehouse` wh
            ON wh.name = bin.warehouse
        WHERE wh.parent_warehouse = 'Wholesale - AGSB'
            AND bin.actual_qty > 0 
        GROUP BY bin.warehouse, item.purity
        HAVING SUM(bin.actual_qty) > 0
    """, as_dict=True)

    bags = {}

    for r in rows:
        bag = bags.setdefault(r.bag_id, {
            "bag_id": r.bag_id,
            "total_weight": 0,
            "avg_purity": 0,
            "total_cost": 0,
            "cost_per_gram": 0,
            "xau_g": 0,
            "purities": []
        })

        purity = flt(r.purity)
        weight = flt(r.weight)
        rate = flt(r.rate)
        amount = flt(r.amount)

        purity_decimal = purity / 1000
        xau = weight * purity_decimal

        bag["purities"].append({
            "purity": purity,
            "weight": weight,
            "rate": rate,
            "xau": xau
        })

        bag["total_weight"] += weight
        bag["total_cost"] += amount
        bag["xau_g"] += xau

    final_list = []
    for bag in bags.values():
        total_weight = bag["total_weight"]
        total_cost = bag["total_cost"]

        total_purity_weight = sum(p["purity"] * p["weight"] for p in bag["purities"])
        avg_purity = (total_purity_weight / total_weight) if total_weight else 0

        bag["avg_purity"] = round(avg_purity, 2)
        bag["total_weight"] = round(total_weight, 2)
        bag["total_cost"] = round(total_cost, 2)
        bag["cost_per_gram"] = round(total_cost / total_weight, 2) if total_weight else 0
        bag["xau_g"] = round(bag["xau_g"], 3)

        final_list.append(bag)

    return final_list

@frappe.whitelist()
def get_bag_details(bag_id):
    """
    Returns summary + items of selected bag.
    XAU is calculated: weight * (purity / 1000)
    """

    rows = frappe.db.sql("""
        SELECT
            item.purity,
            SUM(bin.actual_qty) AS weight,
            AVG(bin.valuation_rate) AS rate,
            SUM(bin.actual_qty * bin.valuation_rate) AS amount
        FROM `tabBin` bin
        JOIN `tabItem` item ON item.name = bin.item_code
        WHERE bin.warehouse = %s
          AND bin.actual_qty > 0
        GROUP BY item.purity
    """, (bag_id,), as_dict=True)

    if not rows:
        return {
            "summary": {},
            "items": []
        }

    total_weight = sum(flt(r.weight) for r in rows)
    total_cost = sum(flt(r.amount) for r in rows)

    total_purity_weight = sum(flt(r.purity) * flt(r.weight) for r in rows)
    avg_purity = (total_purity_weight / total_weight) if total_weight else 0

    # calculate total XAU
    total_xau = sum(flt(r.weight) * (flt(r.purity) / 1000) for r in rows)

    summary = {
        "source_bag": bag_id,
        "total_weight_g": round(total_weight, 2),
        "pure_gold_xau_g": round(total_xau, 3),
        "average_purity": round(avg_purity, 2),
        "total_cost_basis": round(total_cost, 2),
        "cost_per_gram": round(total_cost / total_weight, 2) if total_weight else 0,

        # UI fields (uncalculated here)
        "record_id": "",
        "record_date": ""
    }

    items = []
    for r in rows:
        purity = flt(r.purity)
        weight = flt(r.weight)
        rate = flt(r.rate)
        amount = flt(r.amount)
        xau = weight * (purity / 1000)

        items.append({
            "purity": purity,
            "weight_g": weight,
            "xau_g": round(xau, 3),
            "cost_rm": amount,
            "cost_per_g_rm": rate
        })

    return {
        "summary": summary,
        "items": items
    }
    
@frappe.whitelist()
def save_melt_assay_sales(payload):
	"""
	Create or Update Melt and Assay Sales document
	- If payload.name exists -> update
	- Else -> create new
	"""

	if isinstance(payload, str):
		payload = frappe.parse_json(payload)

	docname = payload.get("name")
	header = payload.get("header", {})
	bag_contents = payload.get("bag_contents", [])
	locked_rates = payload.get("locked_rates", [])

	# ----------------------------------------
	# CREATE OR FETCH DOCUMENT
	# ----------------------------------------
	if docname:
		doc = frappe.get_doc("Melt and Assay Sales", docname)
	else:
		doc = frappe.new_doc("Melt and Assay Sales")
		doc.posting_date = nowdate()

	# ----------------------------------------
	# SET PARENT FIELDS
	# ----------------------------------------
	set_parent_fields(doc, header)

	# ----------------------------------------
	# RESET & APPEND CHILD TABLES
	# ----------------------------------------
	doc.set("bag_contents", [])
	for row in bag_contents:
		doc.append("bag_contents", {
			"purity": row.get("purity"),
			"weight": flt(row.get("weight")),
			"avco": flt(row.get("avco")),
			"cost": flt(row.get("cost")),
			"xau": flt(row.get("xau")),
			"xau_avco": flt(row.get("xau_avco")),
		})

	doc.set("locked_rates", [])
	for row in locked_rates:
		doc.append("locked_rates", {
			"price_per_xau": flt(row.get("price_per_xau")),
			"xau_weight": flt(row.get("xau_weight")),
			"amount": flt(row.get("amount")),
			"remark": row.get("remark"),
		})

	# ----------------------------------------
	# SAVE DOCUMENT
	# ----------------------------------------
	doc.save(ignore_permissions=True)
	frappe.db.commit()

	return {
		"name": doc.name,
		"message": "Melt and Assay Sales saved successfully"
	}

def set_parent_fields(doc, data):
	"""
	Map all parent fields from UI payload to DocType
	"""

	parent_fields = [
		# Bag Summary
		"total_weight", "avg_purity", "total_xau", "total_cost", "xau_avco",

		# Melting
		"weight_before_melting", "weight_after_melting", "melting_cost",
		"melting_payment_mode", "weight_loss", "xau_loss", "loss_percentage",

		# Assay
		"current_avg_purity", "assay_purity", "purity_variance",
		"xau_weight_variance", "actual_xau_weight", "assay_sample_weight",
		"net_xau_sellable", "assay_cost", "assay_payment_mode",

		# Sales
		"sale_net_weight", "sale_assay_purity", "sale_net_xau",
		"total_xau_sold", "total_revenue", "weighted_avg_rate",

		# Metrics - Weight & Purity
		"m_original_gross_weight", "m_weight_after_melting", "m_weight_loss",
		"m_weight_loss_percentage", "m_xau_weight_loss", "m_net_weight_sale",
		"m_original_avg_purity", "m_assay_purity", "m_purity_variance",
		"m_xau_weight_variance",

		# Metrics - Cost
		"m_original_gold_cost", "m_melting_cost", "m_assay_cost",
		"m_total_cost",

		# Metrics - Revenue
		"m_total_revenue", "m_total_cost_profit",
		"m_gross_profit", "m_profit_margin",

		# Metrics - Efficiency
		"m_melting_efficiency", "m_xau_recovery",
		"m_net_sellable", "m_profit_per_xau",

		# Vs Last Sale
		"vs_weight_loss_percentage", "vs_xau_recovery_rate",
		"vs_purity_variance", "vs_net_sellable_percentage",
		"vs_profit_margin",
	]

	for field in parent_fields:
		if field in data:
			doc.set(field, data.get(field))

@frappe.whitelist()
def get_resume_data(log_id):
    """
    Return a JSON payload that contains all data needed to fully
    restore a saved Melt and Assay Sales document on the frontend.

    Response shape:
    {
      "name": doc.name,
      "header": { ... parent fields ... },
      "bag_contents": [ { purity, weight, avco, cost, xau, xau_avco }, ... ],
      "locked_rates": [ { price_per_xau, xau_weight, amount, remark }, ... ]
    }
    """
    if not log_id:
        frappe.throw(_("log_id is required"))

    try:
        doc = frappe.get_doc("Melt and Assay Sales", log_id)
    except Exception as e:
        frappe.throw(_("Cannot load document: {0}").format(str(e)))

    # Build header dict — include all parent fields used by frontend
    header = {}
    parent_fields = [
        # Bag Summary
        "total_weight", "avg_purity", "total_xau", "total_cost", "xau_avco",

        # Melting
        "weight_before_melting", "weight_after_melting", "melting_cost",
        "melting_payment_mode", "weight_loss", "xau_loss", "loss_percentage",

        # Assay
        "current_avg_purity", "assay_purity", "purity_variance",
        "xau_weight_variance", "actual_xau_weight", "assay_sample_weight",
        "net_xau_sellable", "assay_cost", "assay_payment_mode",

        # Sales
        "sale_net_weight", "sale_assay_purity", "sale_net_xau",
        "total_xau_sold", "total_revenue", "weighted_avg_rate",

        # Metrics
        "m_original_gross_weight", "m_weight_after_melting", "m_weight_loss",
        "m_weight_loss_percentage", "m_xau_weight_loss", "m_net_weight_sale",
        "m_original_avg_purity", "m_assay_purity", "m_purity_variance",
        "m_xau_weight_variance", "m_original_gold_cost", "m_melting_cost",
        "m_assay_cost", "m_total_cost", "m_total_revenue", "m_total_cost_profit",
        "m_gross_profit", "m_profit_margin", "m_melting_efficiency",
        "m_xau_recovery", "m_net_sellable", "m_profit_per_xau",

        # Vs last sale
        "vs_weight_loss_percentage", "vs_xau_recovery_rate", "vs_purity_variance",
        "vs_net_sellable_percentage", "vs_profit_margin",
    ]

    for f in parent_fields:
        header[f] = doc.get(f)

    # bag_contents child table
    bag_contents = []
    for row in (doc.get("bag_contents") or []):
        bag_contents.append({
            "purity": row.get("purity"),
            "weight": row.get("weight"),
            "avco": row.get("avco"),
            "cost": row.get("cost"),
            "xau": row.get("xau"),
            "xau_avco": row.get("xau_avco"),
        })

    # locked_rates child table
    locked_rates = []
    for row in (doc.get("locked_rates") or []):
        locked_rates.append({
            "price_per_xau": row.get("price_per_xau"),
            "xau_weight": row.get("xau_weight"),
            "amount": row.get("amount"),
            "remark": row.get("remark"),
        })

    return {
        "name": doc.name,
        "header": header,
        "bag_contents": bag_contents,
        "locked_rates": locked_rates,
    }
